### DISCUSSION TITLE
trailing slash at directory

### DISCUSSION THREAD

**Discussion by Lex-2008** [pocketbase/pocketbase#96] [Show and tell]
at https://github.com/labstack/echo/issues/2211, @ganigeorgiev wrote:
> there seems to be a convention that a directory path should end with `/`.
> I don't know why this is a thing, for me `/dir` and `/dir/` should result in the same response.

Good question!

I believe, it's done primarily for resolving relative links in browsers. If user requested a `/page`, and server replies with HTML with some link in it - say, `<img src="image">` - then browser doesn't know if `/page` is actually a directory (and image should be requested at `/page/image`) or a separate file (and image should be requested at `/image`).

Slash at the end of directory lets browser cleanup this ambiguity.

FWIW, I also saw a similar thing with symlinks on Linux:
```
[lex@flower2 ~]$ ls -l .wine
lrwxrwxrwx 1 lex lex 26 Apr  7  2020 .wine -> /mnt/d/wine
```
compare it with:
```
[lex@flower2 ~]$ ls -l .wine/
total 3292
drwxrwxr-x  2 lex lex    4096 Jul 10 16:43 dosdevices
drwxrwxr-x  7 lex lex    4096 Apr  7  2020 drive_c
...
```

Also rsync distinguishes between "copying `directory` as a whole" and "copying whole contents of a `directory/`".

**Comment by ganigeorgiev** [comment]
> I believe, it's done primarily for resolving relative links in browsers. If user requested a /page, and server replies with HTML with some link in it - say, `<img src="image">` - then browser doesn't know if /page is actually a directory (and image should be requested at /page/image) or a separate file (and image should be requested at /image).

The browsers, when used with http/https schema, doesn't really know about directories. It is just a path. But you are correct about the nuances how relative urls are resolved.

In addition to that, I think web servers have this convention to differentiate the directory listing address (when enabled) from the `index.html` serving one, aka.:
```
- /docs   - show directory listing if enabled
- /docs/  - will look for index.html
```
But the above is rare for production servers.

In any case, in PocketBase there is a default `RemoveTrailingSlash` middleware that could conflicts when serving static content, since it doesn't know when `/page` is a static directory handler or dynamic action and therefore I've created this small helper method - https://github.com/pocketbase/pocketbase/blob/master/apis/base.go#L107-L127.

In future, I'll consider adding more configurable options (similar to `DefaultDataDir`, `ShowStartBanner`, `DefaultDebug`, etc.) so that user could toggle on/off the default middlewares when they need a custom behavior.
